Upgrade GraalVM to 19.3.0

gwenneg opened the pull request:
This PR is related to #4218

It is incomplete and not meant to be merged for now. The build will most probably fail until the GraalVM 19.3.0 dependency is available and integrated into Quarkus.

gwenneg committed:
Upgrade GraalVM to 19.3.0

geoand commented:
How does this relate to #5353 ?

gwenneg commented:
This PR could be considered as a follow-up of #5353.

gwenneg commented:
I'm changing this PR a bit since GraalVM `19.3.0` was released today.

@cescoffier Could you please create the `ubi-quarkus-native-image` for `19.3.0`?

I need to add CI jobs to run JDK 11 native tests before this PR can be promoted. #5353 must also be merged before this one.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -299,10 +299,10 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
         final String vmName = System.getProperty("java.vm.name");
         log.info("Running Quarkus native-image plugin on " + vmName);
-        final List<String> obsoleteGraalVmVersions = Arrays.asList("1.0.0", "19.0.", "19.1.", "19.2.0");
+        final List<String> obsoleteGraalVmVersions = Arrays.asList("1.0.0", "19.0.", "19.1.", "19.2.0", "19.2.1");
```
just have a `19.2.` entry.

gwenneg replied:
Done, thanks for catching this.

gsmet commented on bom/runtime/pom.xml:
```diff
@@ -74,7 +74,7 @@
         <plexus-component-annotations.version>1.7.1</plexus-component-annotations.version>
         <!-- What we actually depend on for the annotations, as latest Graal is not available in Maven fast enough: -->
-        <graal-sdk.version>19.2.1</graal-sdk.version>
+        <graal-sdk.version>19.3.0</graal-sdk.version>
```
Apparently they changed the groupId of the artifacts. At least that was in the announcement.

gwenneg replied:
Fixed. I didn't see that, thanks for the info!

gsmet reviewed (changes requested):
Yeah, don't rush on adding the native image JDK 11 tests. I think we need to define our strategy regarding JDK 11 support and I would probably make it a separate PR once this one is in.

gwenneg commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -176,8 +176,6 @@ public NativeImageBuildItem build(NativeConfig nativeConfig, NativeImageSourceJa
             command.add("-jar");
             command.add(runnerJarName);
-            //https://github.com/oracle/graal/issues/660
-            command.add("-J-Djava.util.concurrent.ForkJoinPool.common.parallelism=1");
```
Fixes #5631

gwenneg commented:
PR promoted and ready to be reviewed/merged as soon as the `ubi-quarkus-native-image` is available for `19.3.0`.

cescoffier commented:
Starting working on the docker image, I keep you posted.

gwenneg commented:
A lot of native builds fail with an error I never saw before:
```
Status: Downloaded newer image for quay.io/quarkus/ubi-quarkus-native-image:19.3.0-java8
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] docker run -v /home/vsts/work/1/s/integration-tests/jpa-derby/target/quarkus-integration-test-jpa-derby-999-SNAPSHOT-native-image-source-jar:/project:z --user 1001:117 --rm quay.io/quarkus/ubi-quarkus-native-image:19.3.0-java8 -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-DCoordinatorEnvironmentBean.transactionStatusManagerEnable=false -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -J-Xmx6g -H:-AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services -H:+JNI --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]    classlist:  13,753.56 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]        (cap):   1,117.15 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]        setup:   3,138.49 ms
22:00:05,682 INFO  [org.hib.Version] HHH000412: Hibernate Core {5.4.9.Final}
22:00:05,736 INFO  [org.hib.ann.com.Version] HCANN000001: Hibernate Commons Annotations {5.1.0.Final}
22:00:05,811 INFO  [org.hib.dia.Dialect] HHH000400: Using dialect: org.hibernate.dialect.DerbyTenSevenDialect
22:00:09,234 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Final
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]   (typeflow):  67,807.49 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]    (objects):  27,713.14 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]   (features):   1,126.19 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]     analysis:  98,826.18 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]     (clinit):   1,270.38 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]     universe:   3,978.92 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]      (parse):  11,151.25 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]     (inline):  17,617.43 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]    (compile):  77,386.47 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]      compile: 109,618.66 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]        image:   4,547.37 ms
[quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner:20]        write:   1,401.82 ms
Fatal error: java.lang.RuntimeException: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /project/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner -z noexecstack -Wl,--gc-sections -Wl,--dynamic-list -Wl,/tmp/SVM-7977477395200125915/exported_symbols.list -Wl,-x -L/tmp/SVM-7977477395200125915 -L/opt/graalvm/jre/lib -L/opt/graalvm/jre/lib/svm/clibraries/linux-amd64 /tmp/SVM-7977477395200125915/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner.o /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm/jre/lib/libnet.a /opt/graalvm/jre/lib/libzip.a /opt/graalvm/jre/lib/libjava.a /opt/graalvm/jre/lib/libnio.a /opt/graalvm/jre/lib/libsunec.a /opt/graalvm/jre/lib/libsunec.a -lpthread -ldl -lpthread -lz -lm -lrt -lstdc++

Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --disable-libmpx --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 8.3.1 20190507 (Red Hat 8.3.1-4) (GCC) 
COMPILER_PATH=/usr/libexec/gcc/x86_64-redhat-linux/8/:/usr/libexec/gcc/x86_64-redhat-linux/8/:/usr/libexec/gcc/x86_64-redhat-linux/:/usr/lib/gcc/x86_64-redhat-linux/8/:/usr/lib/gcc/x86_64-redhat-linux/
LIBRARY_PATH=/usr/lib/gcc/x86_64-redhat-linux/8/:/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/usr/lib/gcc/x86_64-redhat-linux/8/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-o' '/project/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner' '-z' 'noexecstack' '-L/tmp/SVM-7977477395200125915' '-L/opt/graalvm/jre/lib' '-L/opt/graalvm/jre/lib/svm/clibraries/linux-amd64' '-mtune=generic' '-march=x86-64'
 /usr/libexec/gcc/x86_64-redhat-linux/8/collect2 -plugin /usr/libexec/gcc/x86_64-redhat-linux/8/liblto_plugin.so -plugin-opt=/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper -plugin-opt=-fresolution=/tmp/ccJ2IHwq.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --no-add-needed --eh-frame-hdr --hash-style=gnu -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o /project/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner -z noexecstack /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/8/crtbegin.o -L/tmp/SVM-7977477395200125915 -L/opt/graalvm/jre/lib -L/opt/graalvm/jre/lib/svm/clibraries/linux-amd64 -L/usr/lib/gcc/x86_64-redhat-linux/8 -L/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux/8/../../.. --gc-sections --dynamic-list /tmp/SVM-7977477395200125915/exported_symbols.list -x /tmp/SVM-7977477395200125915/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner.o /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm/jre/lib/libnet.a /opt/graalvm/jre/lib/libzip.a /opt/graalvm/jre/lib/libjava.a /opt/graalvm/jre/lib/libnio.a /opt/graalvm/jre/lib/libsunec.a /opt/graalvm/jre/lib/libsunec.a -lpthread -ldl -lpthread -lz -lm -lrt -lstdc++ -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-redhat-linux/8/crtend.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crtn.o
/usr/bin/ld: cannot find -lstdc++
collect2: error: ld returned 1 exit status

	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:462)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:315)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:454)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:115)
Caused by: java.lang.RuntimeException: host C compiler or linker does not seem to work: java.lang.RuntimeException: returned 1

Running command: cc -v -o /project/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner -z noexecstack -Wl,--gc-sections -Wl,--dynamic-list -Wl,/tmp/SVM-7977477395200125915/exported_symbols.list -Wl,-x -L/tmp/SVM-7977477395200125915 -L/opt/graalvm/jre/lib -L/opt/graalvm/jre/lib/svm/clibraries/linux-amd64 /tmp/SVM-7977477395200125915/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner.o /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm/jre/lib/libnet.a /opt/graalvm/jre/lib/libzip.a /opt/graalvm/jre/lib/libjava.a /opt/graalvm/jre/lib/libnio.a /opt/graalvm/jre/lib/libsunec.a /opt/graalvm/jre/lib/libsunec.a -lpthread -ldl -lpthread -lz -lm -lrt -lstdc++

Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl --disable-libmpx --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 8.3.1 20190507 (Red Hat 8.3.1-4) (GCC) 
COMPILER_PATH=/usr/libexec/gcc/x86_64-redhat-linux/8/:/usr/libexec/gcc/x86_64-redhat-linux/8/:/usr/libexec/gcc/x86_64-redhat-linux/:/usr/lib/gcc/x86_64-redhat-linux/8/:/usr/lib/gcc/x86_64-redhat-linux/
LIBRARY_PATH=/usr/lib/gcc/x86_64-redhat-linux/8/:/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/usr/lib/gcc/x86_64-redhat-linux/8/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-v' '-o' '/project/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner' '-z' 'noexecstack' '-L/tmp/SVM-7977477395200125915' '-L/opt/graalvm/jre/lib' '-L/opt/graalvm/jre/lib/svm/clibraries/linux-amd64' '-mtune=generic' '-march=x86-64'
 /usr/libexec/gcc/x86_64-redhat-linux/8/collect2 -plugin /usr/libexec/gcc/x86_64-redhat-linux/8/liblto_plugin.so -plugin-opt=/usr/libexec/gcc/x86_64-redhat-linux/8/lto-wrapper -plugin-opt=-fresolution=/tmp/ccJ2IHwq.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --no-add-needed --eh-frame-hdr --hash-style=gnu -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o /project/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner -z noexecstack /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crt1.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crti.o /usr/lib/gcc/x86_64-redhat-linux/8/crtbegin.o -L/tmp/SVM-7977477395200125915 -L/opt/graalvm/jre/lib -L/opt/graalvm/jre/lib/svm/clibraries/linux-amd64 -L/usr/lib/gcc/x86_64-redhat-linux/8 -L/usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/usr/lib/gcc/x86_64-redhat-linux/8/../../.. --gc-sections --dynamic-list /tmp/SVM-7977477395200125915/exported_symbols.list -x /tmp/SVM-7977477395200125915/quarkus-integration-test-jpa-derby-999-SNAPSHOT-runner.o /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libffi.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/liblibchelper.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libjvm.a /opt/graalvm/jre/lib/svm/clibraries/linux-amd64/libstrictmath.a /opt/graalvm/jre/lib/libnet.a /opt/graalvm/jre/lib/libzip.a /opt/graalvm/jre/lib/libjava.a /opt/graalvm/jre/lib/libnio.a /opt/graalvm/jre/lib/libsunec.a /opt/graalvm/jre/lib/libsunec.a -lpthread -ldl -lpthread -lz -lm -lrt -lstdc++ -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-redhat-linux/8/crtend.o /usr/lib/gcc/x86_64-redhat-linux/8/../../../../lib64/crtn.o
/usr/bin/ld: cannot find -lstdc++
collect2: error: ld returned 1 exit status

	at com.oracle.svm.hosted.image.NativeBootImageViaCC.write(NativeBootImageViaCC.java:382)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:652)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Image build request failed with exit status 1
```

gwenneg commented:
@dmlloyd @bobmcwhirter Would you have an idea of what's causing the error above?

bobmcwhirter commented:
Maybe needs g++ installed?

gwenneg commented:
Thanks @bobmcwhirter! I suspected an issue with the `ubi-quarkus-native-image` but I wasn't sure.

@cescoffier I'll probably need your help to fix that.

cescoffier commented:
@gwenneg Let me try to add the `gcc-c++` package. It was not required before...

gwenneg commented:
Thanks @cescoffier, the new `ubi-quarkus-native-image` image solved the C++ issue.

Most native builds are green now, there are a few unrelated errors left that I'm analyzing now.

gwenneg commented:
Three of the native builds are currently failing because of a NPE looking like the following one. This is happening during the security services registration. Since the exception didn't give any details about the service that was causing it, I added some traces to GraalVM and recompiled it. The traces revealed that the exception is being caused by:
- `org.wildfly.security.password.impl.PasswordFactorySpiImpl` for the `elytron-security-oauth2` extension native tests 
- `org.bouncycastle.jce.provider.X509StoreCertCollection` for the `keycloak-authorization` and `jgit` extensions native tests

Here's one of the exceptions:

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /Library/Java/JavaVirtualMachines/graalvm-unknown-java8-19.3.0/Contents/Home/jre/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar quarkus-integration-test-elytron-security-oauth2-999-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:EnableURLProtocols=http,https --enable-all-security-services -H:+JNI --no-server -H:-UseServiceLoaderFeature -H:+StackTrace quarkus-integration-test-elytron-security-oauth2-999-SNAPSHOT-runner
[quarkus-integration-test-elytron-security-oauth2-999-SNAPSHOT-runner:25596]    classlist:  22,298.82 ms
[quarkus-integration-test-elytron-security-oauth2-999-SNAPSHOT-runner:25596]        (cap):   7,631.83 ms
[quarkus-integration-test-elytron-security-oauth2-999-SNAPSHOT-runner:25596]        setup:  13,156.41 ms
23:43:56,110 INFO  [org.jbo.threads] JBoss Threads version 3.0.0.Final
[quarkus-integration-test-elytron-security-oauth2-999-SNAPSHOT-runner:25596]     analysis:  76,012.36 ms
Fatal error: java.lang.NullPointerException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598)
	at java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1005)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:462)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:315)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:454)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:115)
Caused by: java.lang.NullPointerException
	at sun.reflect.UnsafeFieldAccessorImpl.ensureObj(UnsafeFieldAccessorImpl.java:57)
	at sun.reflect.UnsafeQualifiedObjectFieldAccessorImpl.get(UnsafeQualifiedObjectFieldAccessorImpl.java:38)
	at java.lang.reflect.Field.get(Field.java:393)
	at com.oracle.svm.hosted.SecurityServicesFeature.lambda$getConsParamClassAccessor$0(SecurityServicesFeature.java:279)
	at com.oracle.svm.hosted.SecurityServicesFeature.register(SecurityServicesFeature.java:313)
	at com.oracle.svm.hosted.SecurityServicesFeature.registerServicesForReflection(SecurityServicesFeature.java:191)
	at com.oracle.svm.hosted.ReachabilityHandlerFeature.duringAnalysis(ReachabilityHandlerFeature.java:114)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$8(NativeImageGenerator.java:710)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:63)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:710)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:530)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
	at java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1386)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Error: Image build request failed with exit status 1
```

Do not hesitate to add a comment here if you have an idea of a way to fix that. I was unsuccessful so far, still trying...

cc @dmlloyd

gwenneg commented:
I tried to force the registration of the `org.wildfly.security.password.WildFlyElytronPasswordProvider` and  `org.bouncycastle.jce.provider.BouncyCastleProvider` security providers in multiple ways without being successful.

gwenneg commented:
The `-H:+TraceSecurityServices` command-line argument actually shows many BouncyCastle service being registered, so I don't think there's an issue with the provider registration:

```
[...]
Service registered: Provider = BC , Type = KeyStore , Algorithm = PKCS12-DEF , Class = org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi$DefPKCS12KeyStore
Service registered: Provider = BC , Type = KeyStore , Algorithm = PKCS12-3DES-40RC2 , Class = org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi$BCPKCS12KeyStore
Service registered: Provider = BC , Type = KeyStore , Algorithm = PKCS12-3DES-3DES , Class = org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi$BCPKCS12KeyStore3DES
Service registered: Provider = BC , Type = KeyStore , Algorithm = PKCS12-DEF-3DES-40RC2 , Class = org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi$DefPKCS12KeyStore
Service registered: Provider = BC , Type = KeyStore , Algorithm = PKCS12-DEF-3DES-3DES , Class = org.bouncycastle.jcajce.provider.keystore.pkcs12.PKCS12KeyStoreSpi$DefPKCS12KeyStore3DES
Service registered: Provider = BC , Type = SecureRandom , Algorithm = DEFAULT , Class = org.bouncycastle.jcajce.provider.drbg.DRBG$Default
Service registered: Provider = BC , Type = SecureRandom , Algorithm = NONCEANDIV , Class = org.bouncycastle.jcajce.provider.drbg.DRBG$NonceAndIV
```

Then it fails with `org.bouncycastle.jce.provider.X509StoreCertCollection`.

gwenneg commented:
@cescoffier There's something weird in our CI about the GraalVM version:

`[ERROR] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Out of date build of GraalVM detected: snapshot! Please upgrade to GraalVM 19.3.0.`

This line is created using the GraalVM [Version](https://github.com/oracle/graal/blob/release/graal-vm/19.3/sdk/src/org.graalvm.home/src/org/graalvm/home/Version.java) API which was introduced in 19.3.0. Is this API bugged somehow or could our Docker image be based on `snapshot` rather than `19.3.0`?

gsmet commented:
@gwenneg I think you will need to try to debug what it's trying to do there:
```
Caused by: java.lang.NullPointerException
	at sun.reflect.UnsafeFieldAccessorImpl.ensureObj(UnsafeFieldAccessorImpl.java:57)
	at sun.reflect.UnsafeQualifiedObjectFieldAccessorImpl.get(UnsafeQualifiedObjectFieldAccessorImpl.java:38)
	at java.lang.reflect.Field.get(Field.java:393)
	at com.oracle.svm.hosted.SecurityServicesFeature.lambda$getConsParamClassAccessor$0(SecurityServicesFeature.java:279)
```

gwenneg commented:
@gsmet I did that already and found an issue in GraalVM (see https://github.com/oracle/graal/issues/1883). There's also a significant difference between the JDK and GraalVM in terms of how an unknown security provider is treated and I'm wondering is that could be the cause of our problem here. The NPE wouldn't happen with the JDK. I need to investigate this more and I'll talk about it in the GraalVM ticket then.

machi1990 commented:
> @gsmet I did that already and found an issue in GraalVM (see [oracle/graal#1883](https://github.com/oracle/graal/issues/1883)). 

I just saw this PR https://github.com/oracle/graal/pull/1893 by @jaikiran is that expected to fix the issue?

gwenneg commented:
@machi1990: Yes, @jaikiran added a comment about the PR in https://github.com/oracle/graal/issues/1883. It will certainly help and hopefully fix the issue mentioned here.

machi1990 commented:
> @machi1990: Yes, @jaikiran added a comment about the PR in [oracle/graal#1883](https://github.com/oracle/graal/issues/1883). It will certainly help and hopefully fix the failing Quarkus CI builds (if it does, the only remaining question is when).

Hope to see it for the 19.3.1 release.

gwenneg commented:
I just ran the three failing native builds from the GraalVM 19.3.0 tag (https://github.com/oracle/graal/commit/72d10ce1cd95b094d371e308e922a5960d8c35a8) with a cherry-pick of @jaikiran's fix for the NPE (https://github.com/oracle/graal/pull/1893/commits/dada093fe2b879ae82a8d79c2136084f0a6d4544). This commit in GraalVM solved the Quarkus issues. Thanks again @jaikiran!

The `jgit` and `keycloak-authorization` tests failed at a later native image generation stage because of forbidden instances in the heap, but that's something easy to fix.

gwenneg commented:
I just pushed a commit which substitutes the faulty GraalVM code to include @jaikiran's fix into 19.3.0. It worked locally, let's see if the CI is also happy with is. If that works, it would allow us to use GraalVM 19.3.0 which is kinda comprised for now. Two native builds will fail for another reason as I explained in my previous comment, I'll fix that once I'm sure CI is happy.

gwenneg commented:
It works on my machine ¯\\_(ツ)_/¯

There's a difference that makes this work locally but not with CI... I'll investigate this again tonight.

machi1990 commented:
> It works on my machine ¯\_(ツ)_/¯

Only if it was possible to push your machine 😆 😆

gwenneg commented:
It seems the difference between my machine and the CI came from an outdated `native-image` locally. I made a clean GraalVM install from scratch on my machine and now, the substitution doesn't work (not cool) but it behaves the same locally and on CI (that's better). I asked the GraalVM team on Slack about the possibility to substitute GraalVM code, still waiting for an answer about that.

gsmet commented:
@gwenneg so my current understanding is that the issue Jaikiran fixed is blocking and that we can't update to GraalVM 19.3.0? Is my understanding correct?

@dmlloyd do you have any idea on how we could work around this issue?

dmlloyd commented:
What, in summary, is the issue we're talking about?

gsmet commented:
@dmlloyd Jaikiran pushed this fix into GraalVM: https://github.com/oracle/graal/commit/dada093fe2b879ae82a8d79c2136084f0a6d4544 to solve https://github.com/oracle/graal/issues/1883 which is blocking for us AFAIK.

So either we find some sort of workaround on our side or we will need to wait for 19.3.1.

dmlloyd commented:
I don't see any obvious workaround, sorry.

gwenneg commented:
@gsmet I'm afraid that's correct. I'll still try a few more things tonight about GraalVM code substitution. I saw it work (and I'm 100% sure about it) with a wrong `native-image` version so I'm not ready to give up on this yet :)

jaikiran commented:
Hello @gwenneg, Sorry I am just catching up on this now. To get around the NPE, before 19.3.1 is released, I see that you are trying to `@Substitute` a class that is a `Feature` in Graal. I don't know if substitutions on such classes is supported - I couldn't find any documentation and haven't yet checked the substratevm code to verify.
If you do want to give it a try, then I think what you could attempt is something like this:

```
...
import com.oracle.svm.core.util.VMError;
import com.oracle.svm.util.ReflectionUtil;
...

@TargetClass(className = "com.oracle.svm.hosted.SecurityServicesFeature")
final class SubstituteSecurityServicesFeature {
    
    // Workaround for https://github.com/oracle/graal/issues/1883
    @Substitute
    private static Function<String, Class<?>> getConsParamClassAccessor(BeforeAnalysisAccess access) {
        Map<String, /* EngineDescription */ Object> knownEngines = ReflectionUtil.readStaticField(Provider.class, "knownEngines");
        Field consParamClassNameField = ReflectionUtil.lookupField(access.findClassByName("java.security.Provider$EngineDescription"), "constructorParameterClassName");

        return (serviceType) -> {
            try {
                /* EngineDescription */Object engineDescription = knownEngines.get(serviceType);
                if (engineDescription == null) {
                    return null;
                }
                String constrParamClassName = (String) consParamClassNameField.get(engineDescription);
                if (constrParamClassName != null) {
                    return access.findClassByName(constrParamClassName);
                }
            } catch (IllegalAccessException e) {
                VMError.shouldNotReachHere(e);
            }
            return null;
        };
    }

}
```
Notice that the substitution is on `getConsParamClassAccessor` method and the method impl now has a null check for `engineDescription` to prevent the NPE.

gwenneg commented:
Hi @jaikiran and thanks for your suggestion! That's exactly the kind of thing I've been trying to do.

It actually worked at some point on my machine with a wrong GraalVM installation (I was still using `native-image` from 19.2.1). Here's the code that worked, it's not exactly the same as the one you suggested: https://github.com/quarkusio/quarkus/blob/c75855891ed4e83e2cf2f38eeca13bd24813da06/core/runtime/src/main/java/io/quarkus/runtime/graal/SecurityServicesFeatureSubstitutions.java

I tried to do something very similar to your suggestion at first but GraalVM wasn't happy because the substitution contained a lambda. Anyway, no matter which substitution I use with 19.3.0, it no longer works. It looks like substituting features was possible in 19.2.1 and is no longer possible now.

stuartwdouglas commented:
I am going to have a look at this tomorrow. Now that the native image is built from its own jar it is theoretically possible to modify the native image build to remove these problematic classes.

gsmet commented:
@gwenneg @stuartwdouglas is going to see if we can simply get rid of these two classes somehow.

For the record, they are `org.wildfly.security.password.impl.PasswordFactorySpiImpl` and `org.bouncycastle.jce.provider.X509StoreCertCollection`.

gwenneg commented:
If we remove the problematic classes from the native image, isn't there still a chance this issue could happen with a third-party security provider introduced by some user code? But maybe the risk is so low that it is acceptable considering we would be able to use GraalVM 19.3.0.

dmlloyd commented:
I guess we could completely disable the feature, and copy the feature code into Quarkus with the fix.  It's not a great solution but it would be short-term only.  And it would really depend on whether the feature accesses any APIs.  Also there is a copyright and license issue to deal with; that piece of the code would have to be GPL+CP (so probably its own module, but then we'd have to ban the module from future releases).

stuartwdouglas committed:
Register the password provider at runtime

This prevents the Graal 19.3.0 NPE when checking the
providers list.

stuartwdouglas commented:
I have fixed the issue with the password provide by simply registering it at runtime rather than static init time. 

How do I reproduce the bouncy castle problem? Looking at the CI failures the rest of them seem to be unrelated to the Graal NPE.

gwenneg commented:
@stuartwdouglas The bouncy castle failure should happen while running a native `quarkus-integration-test-jgit` or `quarkus-integration-test-keycloak-authorization` build and you should see the same NPE as the one you fixed.

stuartwdouglas commented:
Thanks, I will check it out. I actually have a solution that removes the classes but it turns out it was not necessary for the elytron tests.

gwenneg commented:
Great!

I'll clean this PR once you're done as some classes won't be needed anymore (`SecurityServicesFeatureSubstitutions` for sure and maybe the ones related to `PortWatcherSubstitutions`).

There could also be additional `quarkus-integration-test-jgit` or `quarkus-integration-test-keycloak-authorization` native build failures unrelated to the GraalVM NPE. That's what `PortWatcherSubstitutions` was meant to fix in the first place, but it might no longer be needed after the classes removal.

stuartwdouglas committed:
Initialize GpgSigner at runtime to avoid security provider problem

stuartwdouglas committed:
Make Keycloak work on Graal 19.3.0

Remove some problematic substitutions and make
bouncycastle classes initialized at runtime to
work around security provider bug

stuartwdouglas commented:
I needed to remove some of the substitutions to get the keycloak tests working. Basically what was happening is that static init methods were registering providers, that were then causing problems for native image. I just moved all this to runtime, which seems to have fixed it (at least in my local testing). 

BTW @gwenneg I noticed that you mentioned you needed to recompile graal to figure out what was causing the problem. Not sure if you are aware of this but if you build with '-Dquarkus.native.debug-build-process=true' then you can just attach a debugger and set a breakpoint to see what is going on without needing a custom graal build. I figured this issue out with a breakpoint at java.security.Security#insertProviderAt to see where the providers were being registered.

gwenneg commented:
Does that work with the community edition of GraalVM? I thought the enterprise edition was needed for debugging. I'm still new to GraalVM so there's probably many things (like this one) I don't know yet :) Thanks for the tip!

stuartwdouglas commented:
Yea, native-image is just a standard java program (as in the actual build process), you can just attach a debugger as normal and debug the build process itself. This does not let you actually debug the resulting program though.

gwenneg commented on core/deployment/src/main/java/io/quarkus/deployment/ThreadLocalRandomProcessor.java:
```diff
@@ -1,15 +0,0 @@
-    @BuildStep
-    RuntimeReinitializedClassBuildItem registerThreadLocalRandomReinitialize() {
-        // ThreadLocalRandom is bugged currently and doesn't reset the seeder
-        // See https://github.com/oracle/graal/issues/1614 for more details
```
This was fixed in GraalVM `19.3.0`. I'm not sure it's OK to remove the Quarkus workaround now as the issue will be present for someone that uses Quarkus `1.1` with a GraalVM version lower than `19.3.0`. Maybe we should wait a few GraalVM versions before removing it.

gsmet replied:
I would vote for removing it now. Otherwise we will forget.

I will add a prominent note that people should upgrade to 19.3.0.

gwenneg commented:
The Quarkus doc needs to be updated, I'll do it today.

gwenneg commented:
Thanks @stuartwdouglas for your help with the NPE issue!

gwenneg commented on docs/src/main/asciidoc/native-and-ssl.adoc:
````diff
@@ -125,7 +125,7 @@ And build again:
 If you check carefully the native executable build options, you can see that the SSL related options are gone:
 ```
-[INFO] [io.quarkus.creator.phase.nativeimage.NativeImagePhase] /opt/graalvm/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dcom.sun.xml.internal.bind.v2.bytecode.ClassTailor.noOptimize=true -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar rest-client-1.0-SNAPSHOT-runner.jar -J-Djava.util.concurrent.ForkJoinPool.comm ...
+[INFO] [io.quarkus.creator.phase.nativeimage.NativeImagePhase] /opt/graalvm/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dcom.sun.xml.internal.bind.v2.bytecode.ClassTailor.noOptimize=true -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -jar rest-client-1.0-SNAPSHOT-runner.jar -J-Djava.util.concurrent.ForkJoinPool.comm ...
````
Review helper: the only thing that changed here is `-H:-JNI` -> `-H:+JNI`.

gsmet commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:
```diff
@@ -132,7 +132,7 @@
      * The docker image to use to do the image build
      */
-    @ConfigItem(defaultValue = "quay.io/quarkus/ubi-quarkus-native-image:19.2.1")
+    @ConfigItem(defaultValue = "quay.io/quarkus/ubi-quarkus-native-image:19.3.0-java8")
```
If we plan to support JDK 11, we need to document how to do it.

And we should at least try to run the tests manually once to check it's OK.

gsmet reviewed (approved):
Awesome team work, thanks.

I think we need some documentation work:
* document how to use GraalVM JDK 11 tech preview if possible, either with Docker or standard install
* see if the SSL documentation needs some updates. From my understanding, it might not be needed anymore to include the SunEC library. We need to properly check that.

@gwenneg do you think you will have the time to do it soon?

gsmet commented:
@gwenneg I let you the honor to push the Merge button, considering how much effort you put into that one. Have fun :).

gwenneg commented:
I'm running tests with both JDK to check everything's OK now. Please don't merge this PR until I'm done.

gwenneg commented:
The JDK 11 native tests are still running locally, I'll merge the PR tomorrow morning once I've checked the results.

I'll also take care of the documentation tomorrow night with a dedicated PR.

bobmcwhirter commented:
You rock, my friend!

On Wed, Dec 4, 2019 at 7:33 PM Gwenneg Lepage <notifications@github.com>
wrote:

> The JDK 11 native tests are still running locally, I'll merge the PR
> tomorrow morning once I've checked the results.
>
> I'll also take care of the documentation tomorrow night with a dedicated
> PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/pull/5358?email_source=notifications&email_token=AAAD4T7AUPY2YJMEUJKMASTQXBD6NA5CNFSM4JLLNSU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF7BJ6Y#issuecomment-561911035>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAD4TYV3ARF557EKG22LT3QXBD6NANCNFSM4JLLNSUQ>
> .
>

gwenneg commented:
The `quarkus-integration-test-elytron-security` native tests failed twice during the JDK 11 tests execution (this is not happening with JDK 8) with the following exception:
```
2019-12-05 02:42:03,821 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-worker-thread-0) HTTP Request to /jca/decodeRSAKey?pemEncoded=MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA0fMGHBxoIKnfS48Z3WD%2Bvx5%2BULCuoYI3JWcfohS%2FrV6Tv%2FlYKpKp1jpGrVBnR3evrRcKNO%2FkT%2FqJc10mhh1sWC4OIyinhxwAAOWS0bPMrLIXT5CWDHADNy%2B8CoHqCx0teqqDAl%2BlN1fz3o%2BdP7Mtkm8YiraccnWTlAfTjppxghAwVlaPZpoDHx%2BMRA%2BjV0Hef6mfTTo3L4oeW4dQj1d%2Fz5olvmPDiGbURHOsI5wODr3IbFJDV2ATnUcscHlpPLDXRvJD59w%2FfLLX2x7K7JAPkwN08KxGSzyT4OGhRXkkb9%2B4JVO91o%2F9obWOch%2FXqydILle6JGl8X31GOjZyH6Gz5TBrUVYyHlF8648kl58r1IWH9M2DX6gA8c0yqVbgpmW21psqGea5M%2BUXBIYaDIGCTtufhajP8JhvendEJG75dWHvVcq3O9L%2BbKwD5taCMvS9zEqneWny0IHAJYLefatxA7PzJK00Y1xFGIVNTcIO%2FcpUCpxTLprygdy0Q%2FTLq%2BSvqvcz3GhGAo9wmmAOqnvcfadYpnsyq7RHb6ku4zChxepvjrzRfMk65DKcsiaL%2BYvZhu4jOLIJzEaC0TG%2B8CfGPT6FttOEcyLfIKpz3vtpLd51YOW%2BhHfcOiHx675foXgTSFacrCI71a8I%2FDXvXPhbUEa7yFbae2rjDBrbCjGooOsCAwEAAQ%3D%3D failed, error id: 75fad95b-be6c-4c9d-a5d9-f6a88846c610-1: org.jboss.resteasy.spi.UnhandledException: java.security.NoSuchAlgorithmException: RSA KeyFactory not available
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:108)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.lambda$handle$0(VertxRequestHandler.java:73)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:316)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:479)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: java.security.NoSuchAlgorithmException: RSA KeyFactory not available
	at java.security.KeyFactory.<init>(KeyFactory.java:138)
	at java.security.KeyFactory.getInstance(KeyFactory.java:172)
	at io.quarkus.it.elytron.KeyFactoryEndpoint.decodeRSAKey(KeyFactoryEndpoint.java:39)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:151)
	at org.jboss.resteasy.core.MethodInjectorImpl.lambda$invoke$3(MethodInjectorImpl.java:122)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)
	at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:628)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1996)
	at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:110)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:122)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:594)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:468)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:421)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:363)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:423)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:391)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invoke$1(ResourceMethodInvoker.java:365)
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:365)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:477)
	... 15 mor
```

gwenneg commented:
I'm merging this PR as it feels we're ready for GraalVM 19.3.0 with JDK 8.

There are still some issues I need to investigate with JDK 11 though, I will continue this work from #5682.

stuartwdouglas commented:
Thanks for your awesome work on this @gwenneg!

geoand commented:
Great stuff!!!

machi1990 commented:
Awesome work @gwenneg!!
